### PR TITLE
Sharding-aware CG solver

### PR DESCRIPTION
**Pull request by sbiquard**
This adds our own implementation of a conjugate gradient solver.

Comparison with lineax's CG solver:
- Unlike lineax, we record the residual norm at every iteration to monitor convergence. (This is the main motivation for this implementation.)
- Unlike lineax, we only use the standard convergence criterion in the 'b' space (given Ay=b), i.e. we do not check if the solution is still updating. (This may change.)
- Like lineax, we stabilise the residual every 10 (by default) iteration steps to improve numerical stability.
- Unlike lineax, our inner products use a reduction rather than `vdot`, so vectors sharded along their contracting axis are solved directly (`vdot` fails under explicit-axis sharding). Tested on explicit/auto axes and 1D/2D meshes.
- Like lineax, we support reverse-mode autodiff, but it is opt-in via `loop_kind='bounded'` (forward-mode works by default).
- Unlike lineax, we do not check curvature by default (`A` is assumed positive definite). The `negative_curvature` parameter controls this behaviour (default is `ignore`, `error` raises, `truncate` stops the iteration).

A couple of notes on performance:
- Reverse-mode autodiff (`loop_kind='bounded'`) is noticeably slower than the default `lax` loop, because its checkpoint structure runs up to `max_steps` regardless of early convergence. That is why it is opt-in; the default forward-only path is unaffected.
- Checking curvature every iteration (via `equinox.error_if`) turned out to be surprisingly expensive, enough to dominate the per-step cost in some instances, which is why curvature is unchecked by default.

### CODE REVIEW

File: src/furax/linalg/_cg.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+
+    Equivalent to ``furax.tree.dot`` but tolerates inputs sharded along their
+    contracting dimensions: leaf reductions are constrained to fully-replicated
+    output sharding, which makes JAX insert an all-reduce instead of erroring.
+    """
+    replicated = _replicated_sharding()
+
+    def leaf_dot(a: Array, b: Array) -> Array:
+        prod = jnp.conj(a) * b
+        if replicated is None:
```
Comment (ASKabalan) [inline src/furax/linalg/_cg.py]: this is not needed as the sharding of the jnp.sum operation is automatically replicated without any extra work

Comment (ASKabalan) [reply src/furax/linalg/_cg.py]:
```py
import os

os.environ["XLA_FLAGS"] = "--xla_force_host_platform_device_count=4"
os.environ["JAX_PLATFORMS"] = "cpu"

import jax
from jax import lax
import jax.numpy as jnp
from jax.sharding import PartitionSpec as P, AxisType , Mesh , NamedSharding , get_abstract_mesh
mesh = jax.make_mesh((4, 1), ("x", "y"), axis_types=(AxisType.Auto, AxisType.Auto))
sharding = NamedSharding(mesh, P("x", "y"))
rng_key = jax.random.PRNGKey(42)
a = jax.random.normal(rng_key, (16 , 8))
a = lax.with_sharding_constraint(a , sharding)
b = jax.random.normal(rng_key , (16 , 8))
print(f"sharding of a is {a.sharding.spec}")
# Out : sharding of a is P('x', 'y')
b = lax.with_sharding_constraint(b , sharding)
print(f"sharding of b is {b.sharding.spec}")
# Out : sharding of b is P('x', 'y')

@jax.jit
def dot(a , b):
    prod = jnp.conj(a) * b
    return jnp.sum(prod)
dotted = dot(a , b)
print(f"sharding of dotted is {dotted.sharding.spec}")
# Out : sharding of dotted is P()
```

File: src/furax/linalg/_cg.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+    """
+
+    solution: PyTree[Num[Array, '...']]
+    residuals: Float[Array, ' max_steps']
+    num_steps: Array
+
+
+def _replicated_sharding() -> NamedSharding | None:
+    """Return a fully-replicated NamedSharding for the active mesh, or None."""
+    mesh = jax.sharding.get_abstract_mesh()
```
Comment (ASKabalan) [inline src/furax/linalg/_cg.py]: This forces the user to set the mesh using `jax.set_mesh(mesh)` 
In some case we really need the mesh, but in most cases we don't .
I think it shouldn't be used unless we really don't have a choice

File: src/furax/linalg/_cg.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+        prod = jnp.conj(a) * b
+        if replicated is None:
+            return jnp.sum(prod)
+        return jax.lax.with_sharding_constraint(jnp.sum(prod), replicated)  # type: ignore[no-any-return]
+
+    parts = jax.tree.map(leaf_dot, x, y)
+    return sum(jax.tree.leaves(parts), start=jnp.array(0))
+
+
+def _sharded_norm(x: PyTree[Num[Array, '...']]) -> Num[Array, '']:
```
Comment (ASKabalan) [inline src/furax/linalg/_cg.py]: can be inlined

File: src/furax/linalg/_cg.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+    replicated = _replicated_sharding()
+
+    def leaf_dot(a: Array, b: Array) -> Array:
+        prod = jnp.conj(a) * b
+        if replicated is None:
+            return jnp.sum(prod)
+        return jax.lax.with_sharding_constraint(jnp.sum(prod), replicated)  # type: ignore[no-any-return]
+
+    parts = jax.tree.map(leaf_dot, x, y)
+    return sum(jax.tree.leaves(parts), start=jnp.array(0))
```
Comment (ASKabalan) [inline src/furax/linalg/_cg.py]: sharded_dot can also be inlined
leaf dot is basically just `jnp.sum(jnp.conj(a) * b)` which works under sharding.

The only thing that worries my is that a replication might sometimes be problematic for gradient computation (basically losing info of the sharding) but it does not seem to be the case anymore

```py
import os

os.environ["XLA_FLAGS"] = "--xla_force_host_platform_device_count=4"
os.environ["JAX_PLATFORMS"] = "cpu"

import jax
from jax import lax
import jax.numpy as jnp
from jax.sharding import PartitionSpec as P, AxisType , Mesh , NamedSharding , get_abstract_mesh
mesh = jax.make_mesh((4, 1), ("x", "y"), axis_types=(AxisType.Auto, AxisType.Auto))
sharding = NamedSharding(mesh, P("x", "y"))
rng_key = jax.random.PRNGKey(42)
a = jax.random.normal(rng_key, (16 , 8))
a = lax.with_sharding_constraint(a , sharding)
b = jax.random.normal(rng_key , (16 , 8))
print(f"sharding of a is {a.sharding.spec}")
# Out : sharding of a is P('x', 'y')
b = lax.with_sharding_constraint(b , sharding)
print(f"sharding of b is {b.sharding.spec}")
# Out : sharding of b is P('x', 'y')

@jax.jit
def dot(a , b):
    prod = jnp.conj(a) * b
    return jnp.sum(prod)
dotted = dot(a , b)
print(f"sharding of dotted is {dotted.sharding.spec}")
# Out : sharding of dotted is P()

grad_dot = jax.grad(dot, argnums=(0, 1))
grads = grad_dot(a, b)
print(f"sharding of grads[0] is {grads[0].sharding.spec}")
# Out : sharding of grads[0] is P('x', 'y')
print(f"sharding of grads[1] is {grads[1].sharding.spec}")
# Out : sharding of grads[1] is P('x', 'y')
```

File: tests/linalg/test_cg.py [hunk 5]
```diff
@@ -0,0 +1,222 @@
+"""Tests for the Conjugate Gradient solver."""
+
```
Comment (ASKabalan) [inline tests/linalg/test_cg.py:2]: we can also add distributed tests inspired by 

https://github.com/DifferentiableUniverseInitiative/JaxPM/blob/main/tests/test_distributed_pm.py
and the setup in 
https://github.com/DifferentiableUniverseInitiative/JaxPM/blob/main/tests/conftest.py

File: src/furax/linalg/_cg.py [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+        *_, converged, i, _ = carry
+        return ~converged & (i < max_steps)
+
+    def body_fn(carry):  # type: ignore[no-untyped-def]
+        x, r, p, rz, _, i, residuals = carry
+
+        Ap = A(p)
+        pAp = _sharded_dot(p, Ap)
+
+        safe_pAp = jnp.where(pAp == 0, 1.0, pAp)
```
Comment (ASKabalan) [inline src/furax/linalg/_cg.py]: # Negative curvature check
`---> CHECK FOR NEGATIVE CURVATURE HERE <--`
`negative_curvature = pAp <= 0.0`

Negative curvature is typically not allowed for a semi positive definite and the semi negative definite (with a minus sign) 

If we allow it then we should implement a truncated option where the user can basically set truncat=True in this case it will trigger a premature break of the iteration loop

This option can be an enum for example 

MATRIX_TAG:
POSITIVE_SEMI_DEFINITE
NEGATIVE_SEMI_DEFINITE
TRUNACE

You implemented POSITIVE_SEMI_DEFINITE
NEGATIVE_SEMI_DEFINITE is basically the same with double minus signs (one at the beginning to make it POSITIVE_SEMI_DEFINITE and one at the end 

in both cases a negative curvature should cause an error via callback

if truncate then leave early (no error)

ask claude to get inspired to how I did the truncation with the lineax CG and also to get inspired how to do the double minus trick
[cg.py](https://github.com/user-attachments/files/28746995/cg.py)

File: tests/linalg/test_cg.py [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+        """Reverse-mode AD through while_loop is not supported."""
+
+        d = jnp.arange(1.0, 6.0)
+        A = DiagonalOperator(d, in_structure=as_structure(d))
+        b = jnp.ones(5)
+
+        def f(b):
+            return jnp.sum(cg(A, b, max_steps=20).solution)
+
+        with pytest.raises(ValueError, match='Reverse-mode differentiation'):
```
Comment (ASKabalan) [inline tests/linalg/test_cg.py]: we can use reverse mode differentiation (actually reverse mode diff is significantly more important that forward mode since most functions are big input scalar output, basically a likelihood)

We can do that by using the equinox version of a while loop

https://github.com/patrick-kidger/equinox/blob/f16bb5baee1a15031c5c75a08a449cfccdde0c00/equinox/internal/_loop/loop.py#L22


good example (unbiased 😄 )

https://github.com/ASKabalan/jax-fli/blob/main/src/jax_fli/pm/integrate.py

File: src/furax/linalg/_cg.py (hunk 1 above)
Comment (sbiquard) [reply src/furax/linalg/_cg.py]: It also works with explicit axes, thanks good catch

Comment (sbiquard) [reply src/furax/linalg/_cg.py]: It seems the only reason for using some flavour `dot` in this case is that it allows to choose a [Precision](https://docs.jax.dev/en/latest/jax.lax.html#jax.lax.Precision) for the GPU/TPU backend (no effect on CPU), so that reduction remains precise even with float32 inputs (no effect for float64).

File: src/furax/linalg/_cg.py (hunk 2 above)
Comment (sbiquard) [reply src/furax/linalg/_cg.py]: If we use `jnp.sum` then this becomes dead code and we can remove it

**Review by ASKabalan — DISMISSED** [review]
OK very good
This is G2G
Just one comment, the tolerances (1e-4) are not machine precision
In general for an iterative algorithm it is expected to have an error bigger than the float 32 eps (due to compounded error)

```
In [2]: np.finfo(np.float32).eps
Out[2]: np.float32(1.1920929e-07)

In [3]: np.finfo(np.float64).eps
Out[3]: np.float64(2.220446049250313e-16)
```

Not sure if your tests if tests are running in float64 or float32

If float32 then this is fine, but consider running the tests in float64

if in float 64 (which perhaps it is https://github.com/CMBSciPol/furax/blob/sb/cg/tests/conftest.py#L46) 

Then this tolerance might be too relaxed

I approve this PR anyways, I will let you take the decision
